include/experimental/__p0009_bits/mdspan.hpp

```diff
@@ -376,6 +442,23 @@ class mdspan
   MDSPAN_FORCE_INLINE_FUNCTION constexpr mapping_type const& __mapping_ref() const noexcept { return __members.__second().__first(); }
   MDSPAN_FORCE_INLINE_FUNCTION _MDSPAN_CONSTEXPR_14 accessor_type& __accessor_ref() noexcept { return __members.__second().__second(); }
   MDSPAN_FORCE_INLINE_FUNCTION constexpr accessor_type const& __accessor_ref() const noexcept { return __members.__second().__second(); }
+  
+  MDSPAN_TEMPLATE_REQUIRES(
+    class SizeType,
+    /* requires */ (
+      _MDSPAN_TRAIT(std::is_convertible, const SizeType&, index_type) &&
+      _MDSPAN_TRAIT(std::is_nothrow_constructible, index_type, const SizeType&)
+    )
+  )
+  MDSPAN_FORCE_INLINE_FUNCTION constexpr bool __is_index_oor(SizeType index, index_type extent) const noexcept {
+    // Check for negative indices
+    if constexpr(std::is_signed_v<SizeType>) {
+      if(index < 0) {
+        return true;
+      }
+    }
```

Comment:
`SizeType` doesn't need to be integral here or even less-than comparable with `int`.  mdspan behaves as if all operations on indices happen after conversion to `index_type`.  Thus, should we consider instead converting to `index_type` first, as we do below?

```suggestion
    // Check for negative indices
    if constexpr (std::is_signed_v<index_type>) {
      if (static_cast<index_type>(index) < 0) {
        return true;
      }
    }
```

Reply:
Seems reasonable

Reply:
After thinking about it again, I'm not sure if this is correct: imagine I have `unsigned short` as `index_type` and an extent of size of `65535`. Now, I want to access with a `signed short` of value `-10`. What I end up getting is the element at index `65526`, which is not the expected result.

Reply:
I ended up reverting this: if `SizeType` is not integral or less-than comparable with int, `std::is_signed_v` should be false anyway. The question is, what happens if `SizeType` can be negative, but is is not integral or less-than comparable with int, while `index_type` is not signed?

Reply:
> After thinking about it again, I'm not sure if this is correct: imagine I have unsigned short as index_type and an extent of size of 65535. Now, I want to access with a signed short of value -10. What I end up getting is the element at index 65526, which is not the expected result.

Thanks for pointing this out!  The underlying issue, as I understand it, is that _`index-cast`_ returns its input (with the same type, not cast to `index_type`) if the input type is integral-not-bool.

[[mdspan.mdspan.members] 2 - 3](https://eel.is/c++draft/views.multidim#mdspan.mdspan.members-2) together say that the precondition on the input of `mdspan::operator[]` applies to `extents_type::`_`index-cast`_`(std::move(indices...))`.  For the specific case you mentioned of `SizeType = signed short`, [[mdspan.extents.expo] 9](https://eel.is/c++draft/views.multidim#mdspan.extents.expo-9) explains that _`index-cast`_ would just return `signed short`; it does not convert to `index_type`.  -10 cannot be part of a multidimensional index in `extents()`, so this violates a precondition.

Reply:
Btw, I was looking through [extents.hpp](https://github.com/kokkos/mdspan/blob/stable/include/experimental/__p0009_bits/extents.hpp) and noticed `check_all_indices`.  It looks like the layout mappings using this.  I think it could be refactored a bit to serve your needs.